server/plugin/webhook_test.go

```diff
@@ -0,0 +1,127 @@
+package plugin
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/plugin/plugintest"
+	"github.com/mattermost/mattermost/server/public/plugin/plugintest/mock"
+	"github.com/mattermost/mattermost/server/public/pluginapi"
+	"golang.org/x/oauth2"
+
+	"github.com/google/go-github/v54/github"
+	"github.com/stretchr/testify/assert"
+)
+
+const webhookSecret = "whsecret"
+const orgMember = "org-member"
+const orgCollaborator = "org-collaborator"
+const gitHubOrginization = "test-org"
+
+func newPlugin(userID string, gitHubURL string) *Plugin {
+	p := NewPlugin()
+	p.initializeAPI()
+	p.SetDriver(&plugintest.Driver{})
+	p.store = &pluginapi.MemoryStore{}
+	token, _ := generateSecret()
+	encryptionKey, _ := generateSecret()
+	encryptedToken, _ := encrypt([]byte(encryptionKey), token)
+	_, _ = p.store.Set(userID+githubTokenKey, GitHubUserInfo{
+		UserID: userID,
+		Token: &oauth2.Token{
+			AccessToken: encryptedToken,
+		},
+	})
+	p.setConfiguration(&Configuration{
+		EncryptionKey:       encryptionKey,
+		GitHubOrg:           gitHubOrginization,
+		WebhookSecret:       webhookSecret,
+		EnterpriseBaseURL:   gitHubURL,
+		EnterpriseUploadURL: gitHubURL,
+	})
+
+	_ = p.AddSubscription(
+		gitHubOrginization+"/test-repo",
+		&Subscription{
+			ChannelID: "1",
+			CreatorID: userID,
+			Features:  Features(strings.Join([]string{featureIssues, featureIssueCreation}, ",")),
+			Flags:     SubscriptionFlags{IncludeOnlyOrgMembers: true},
+		},
+	)
+
+	return p
+}
+
+func mockGitHubServer(user string) *httptest.Server {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != fmt.Sprintf("/api/v3/orgs/%v/members/%v", gitHubOrginization, user) {
+			http.Error(w, "Not found", http.StatusNotFound)
+			return
+		}
+
+		if user == orgMember {
+			w.WriteHeader(http.StatusNoContent)
+		} else if user == orgCollaborator {
+			w.WriteHeader(http.StatusFound)
+		}
+	}))
+
+	return ts
+}
+
+func TestIncludeOnlyOrgMembers(t *testing.T) {
+	tests := []struct {
+		name         string
+		user         github.User
+		subscription Subscription
+		expectWarn   bool
+		want         bool
+	}{
+		{
+			name: "IncludeOnlyOrgMembers flag is false",
+			user: github.User{
+				Login: github.String(orgMember),
+			},
+			subscription: Subscription{
+				Flags: SubscriptionFlags{IncludeOnlyOrgMembers: false},
+			},
+			expectWarn: false,
+			want:       false,
+		},
+		{
+			name: "Failed to get GitHub Client",
+			user: github.User{
+				Login: github.String(orgMember),
+			},
+			subscription: Subscription{
+				CreatorID: model.NewId(),
+				Flags:     SubscriptionFlags{IncludeOnlyOrgMembers: true},
+			},
+			expectWarn: true,
+			want:       false,
```

Comment:
Thanks for these tests @burakcakirel :+1: Can we have one that tests the happy path of `want: true`?

Reply:
I've tried to test that case but couldn't because the code is unsuitable for testing. I couldn't mock the [`GetGitHubClient`](https://github.com/mattermost/mattermost-plugin-github/blob/d70572bff66222eb32bbdff7f5ee7b7512519d1a/server/plugin/plugin.go#L164) method. Then I've tried to write feature tests for this by mimicking the [Jira Plugin tests](https://github.com/mattermost/mattermost-plugin-jira/blob/36714ad1e75fa0ead5ee30f81b03b8aac540cf5c/server/webhook_http_test.go) but couldn't do it again because [`postIssueEvent`](https://github.com/mattermost/mattermost-plugin-github/blob/d70572bff66222eb32bbdff7f5ee7b7512519d1a/server/plugin/webhook.go#L545) method doesn't return anything like in Jira Plugin's [`PostToChannel`](https://github.com/mattermost/mattermost-plugin-jira/blob/36714ad1e75fa0ead5ee30f81b03b8aac540cf5c/server/webhook.go#L63) method. Hence, this prevents to write a test for this method. Here is the gist file of the tests I've tried to achieve:
https://gist.github.com/burakcakirel/bf8c3da4b67893b70d9871e5032674ae#file-webhook_test-go-L170